TERRAFORM.md

```diff
@@ -0,0 +1,96 @@
+# Terraform Best Practices and Guidelines
+
+General terraform guidelines and best practices
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Introduction](#introduction)
+- [Code Organization](#code-organization)
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
```

Comment:
Maybe we want a nomenclature section or point to the upstream documentation about this. I know I am still sometimes confused about how we use the words "module", "project" and "workspace" which I think is a little different from the way the official Terraform documentation uses them.

Reply:
As I haven't read the terraform docu for a long while maybe could you summarize how you understand it as you have recently read it :smile:
We currently don't use terraform workspace functionality.

Reply:
AFAICT Terraform doesn't have a concept of "project" and a module is just any directory that has .tf files in it. Some modules are "root modules" (i.e. you can run "terraform" on that directory), all the others are "child modules" (or perhaps "orphaned modules" in some cases). Some child modules are reusable. The way I have heard you use these words is a bit different :) I think I can live with any reasonable definition/nomenclature but I would really like we nail one down in this document and my preference it to reuse as much of upstream as possible (and we may still need to define other words/concepts on top of that of course).

https://developer.hashicorp.com/terraform/language/modules

Reply:
I propose to use
- Root Module := for terraform module that we can apply, what I used to call project
- Child Module := for non appliable module, that are tight to a root module (without backend, provider definition) (currently we don't have any of those.)
  - Reusable Module := for module non appliable, that can be used in any root module
@adk-swisstopo what do you think ?

Reply:
Sounds good and I think that matches the upstream definitions. I suggest adding something along the lines of "We use the terms 'module', 'root module' and 'reusable module' in the same way than [the Terraform documentation](https://developer.hashicorp.com/terraform/language/modules)." Maybe also "Historically, we have sometimes used the word 'project' to mean 'root module'."

Reply:
You did that already and I am just confused about the github review workflow. Thank you.